### ISSUE TITLE
system requirements isnt wokring as expected 

### ISSUE THREAD

**Issue by pvbhanuteja** [prefix-dev/pixi#1665]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://github.com/prefix-dev/pixi/releases) of pixi, using `pixi --version`.

### Reproducible example

```
[project]
name = "apple"
version = "0.9.8"
description = "Add a short description here"
authors = [{name = "Newton", email = "apple@gravity.com"}]
requires-python = ">= 3.11"
dependencies = []

[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"

[tool.pixi.project]
channels = ["conda-forge","pytorch"]
platforms = ["linux-64"]

[tool.pixi.pypi-dependencies]
retrieval_api = { path = ".", editable = true }

[tool.pixi.tasks]

[tool.pixi.dependencies]
python = "3.11.*"
fastapi = ">=0.111.1,<0.112"
pytorch = {version = "2.4.0", channel="pytorch"}
```

Above toml lead to install a cuda version of pytorch though I haven't mentioned cuda in system reqs

```
pixi list yielded 
Package               Version     Build                         Size       Kind   Source
_libgcc_mutex         0.1         conda_forge                   2.5 KiB    conda  _libgcc_mutex-0.1-conda_forge.tar.bz2
_openmp_mutex         4.5         2_kmp_llvm                    5.6 KiB    conda  _openmp_mutex-4.5-2_kmp_llvm.tar.bz2
annotated-types       0.7.0       pyhd8ed1ab_0                  17.8 KiB   conda  annotated-types-0.7.0-pyhd8ed1ab_0.conda
anyio                 4.4.0       pyhd8ed1ab_0                  101.8 KiB  conda  anyio-4.4.0-pyhd8ed1ab_0.conda
blas                  1.0         mkl                           1.3 KiB    conda  blas-1.0-mkl.tar.bz2
bzip2                 1.0.8       h4bc722e_7                    246.9 KiB  conda  bzip2-1.0.8-h4bc722e_7.conda
ca-certificates       2024.7.4    hbcca054_0                    151.2 KiB  conda  ca-certificates-2024.7.4-hbcca054_0.conda
certifi               2024.7.4    pyhd8ed1ab_0                  155.6 KiB  conda  certifi-2024.7.4-pyhd8ed1ab_0.conda
click                 8.1.7       unix_pyh707e725_0             82.5 KiB   conda  click-8.1.7-unix_pyh707e725_0.conda
cuda-cudart           12.4.99     hd3aeb46_0                    21.9 KiB   conda  cuda-cudart-12.4.99-hd3aeb46_0.conda
cuda-cudart_linux-64  12.4.99     h59595ed_0                    182.3 KiB  conda  cuda-cudart_linux-64-12.4.99-h59595ed_0.conda
cuda-cupti            12.4.127    he02047a_2                    1.8 MiB    conda  cuda-cupti-12.4.127-he02047a_2.conda
cuda-libraries        12.4.0      ha770c72_0                    19.6 KiB   conda  cuda-libraries-12.4.0-ha770c72_0.conda
cuda-nvrtc            12.4.99     hd3aeb46_0                    18 MiB     conda  cuda-nvrtc-12.4.99-hd3aeb46_0.conda
cuda-nvtx             12.4.127    he02047a_2                    30.9 KiB   conda  cuda-nvtx-12.4.127-he02047a_2.conda
cuda-opencl           12.4.99     h59595ed_0                    29.4 KiB   conda  cuda-opencl-12.4.99-h59595ed_0.conda
cuda-runtime          12.4.0      ha804496_0                    19.5 KiB   conda  cuda-runtime-12.4.0-ha804496_0.conda
cuda-version          12.4        h3060b56_3                    20.5 KiB   conda  cuda-version-12.4-h3060b56_3.conda
dnspython             2.6.1       pyhd8ed1ab_1                  165.5 KiB  conda  dnspython-2.6.1-pyhd8ed1ab_1.conda
email-validator       2.2.0       pyhd8ed1ab_0                  43.1 KiB   conda  email-validator-2.2.0-pyhd8ed1ab_0.conda
email_validator       2.2.0       hd8ed1ab_0                    6.5 KiB    conda  email_validator-2.2.0-hd8ed1ab_0.conda
exceptiongroup        1.2.2       pyhd8ed1ab_0                  19.9 KiB   conda  exceptiongroup-1.2.2-pyhd8ed1ab_0.conda
fastapi               0.111.1     pyhd8ed1ab_0                  69.4 KiB   conda  fastapi-0.111.1-pyhd8ed1ab_0.conda
fastapi-cli           0.0.4       pyhd8ed1ab_0                  14.3 KiB   conda  fastapi-cli-0.0.4-pyhd8ed1ab_0.conda
filelock              3.15.4      pyhd8ed1ab_0                  17.2 KiB   conda  filelock-3.15.4-pyhd8ed1ab_0.conda
gmp                   6.3.0       hac33072_2                    449.3 KiB  conda  gmp-6.3.0-hac33072_2.conda
gmpy2                 2.1.5       py311hc4f1f91_1               200.1 KiB  conda  gmpy2-2.1.5-py311hc4f1f91_1.conda
h11                   0.14.0      pyhd8ed1ab_0                  47.1 KiB   conda  h11-0.14.0-pyhd8ed1ab_0.tar.bz2
h2                    4.1.0       pyhd8ed1ab_0                  45.7 KiB   conda  h2-4.1.0-pyhd8ed1ab_0.tar.bz2
hpack                 4.0.0       pyh9f0ad1d_0                  24.7 KiB   conda  hpack-4.0.0-pyh9f0ad1d_0.tar.bz2
httpcore              1.0.5       pyhd8ed1ab_0                  44.7 KiB   conda  httpcore-1.0.5-pyhd8ed1ab_0.conda
httpx                 0.27.0      pyhd8ed1ab_0                  63.1 KiB   conda  httpx-0.27.0-pyhd8ed1ab_0.conda
hyperframe            6.0.1       pyhd8ed1ab_0                  14.3 KiB   conda  hyperframe-6.0.1-pyhd8ed1ab_0.tar.bz2
icu                   75.1        he02047a_0                    11.6 MiB   conda  icu-75.1-he02047a_0.conda
idna                  3.7         pyhd8ed1ab_0                  51.5 KiB   conda  idna-3.7-pyhd8ed1ab_0.conda
jinja2                3.1.4       pyhd8ed1ab_0                  109 KiB    conda  jinja2-3.1.4-pyhd8ed1ab_0.conda
ld_impl_linux-64      2.40        hf3520f5_7                    691 KiB    conda  ld_impl_linux-64-2.40-hf3520f5_7.conda
libcublas             12.4.2.65   hd3aeb46_0                    228.4 MiB  conda  libcublas-12.4.2.65-hd3aeb46_0.conda
libcufft              11.2.0.44   hd3aeb46_0                    166.9 MiB  conda  libcufft-11.2.0.44-hd3aeb46_0.conda
libcufile             1.9.0.20    hd3aeb46_0                    897.7 KiB  conda  libcufile-1.9.0.20-hd3aeb46_0.conda
libcurand             10.3.5.119  hd3aeb46_0                    39.7 MiB   conda  libcurand-10.3.5.119-hd3aeb46_0.conda
libcusolver           11.6.0.99   hd3aeb46_0                    78.1 MiB   conda  libcusolver-11.6.0.99-hd3aeb46_0.conda
libcusparse           12.3.0.142  hd3aeb46_0                    112.4 MiB  conda  libcusparse-12.3.0.142-hd3aeb46_0.conda
libexpat              2.6.2       h59595ed_0                    72 KiB     conda  libexpat-2.6.2-h59595ed_0.conda
libffi                3.4.2       h7f98852_5                    56.9 KiB   conda  libffi-3.4.2-h7f98852_5.tar.bz2
libgcc-ng             14.1.0      h77fa898_0                    822.4 KiB  conda  libgcc-ng-14.1.0-h77fa898_0.conda
libhwloc              2.11.1      default_hecaa2ac_1000         2.3 MiB    conda  libhwloc-2.11.1-default_hecaa2ac_1000.conda
libiconv              1.17        hd590300_2                    689.2 KiB  conda  libiconv-1.17-hd590300_2.conda
libnpp                12.2.5.2    hd3aeb46_0                    94.3 MiB   conda  libnpp-12.2.5.2-hd3aeb46_0.conda
libnsl                2.0.1       hd590300_0                    32.6 KiB   conda  libnsl-2.0.1-hd590300_0.conda
libnvfatbin           12.4.99     hd3aeb46_0                    773.2 KiB  conda  libnvfatbin-12.4.99-hd3aeb46_0.conda
libnvjitlink          12.4.99     hd3aeb46_0                    15.8 MiB   conda  libnvjitlink-12.4.99-hd3aeb46_0.conda
libnvjpeg             12.3.1.89   h59595ed_0                    2.4 MiB    conda  libnvjpeg-12.3.1.89-h59595ed_0.conda
libsqlite             3.46.0      hde9e2c9_0                    845.1 KiB  conda  libsqlite-3.46.0-hde9e2c9_0.conda
libstdcxx-ng          14.1.0      hc0a3c3a_0                    3.7 MiB    conda  libstdcxx-ng-14.1.0-hc0a3c3a_0.conda
libuuid               2.38.1      h0b41bf4_0                    32.8 KiB   conda  libuuid-2.38.1-h0b41bf4_0.conda
libxcrypt             4.4.36      hd590300_1                    98 KiB     conda  libxcrypt-4.4.36-hd590300_1.conda
libxml2               2.12.7      he7c6b58_4                    690.6 KiB  conda  libxml2-2.12.7-he7c6b58_4.conda
libzlib               1.3.1       h4ab18f5_1                    60.1 KiB   conda  libzlib-1.3.1-h4ab18f5_1.conda
llvm-openmp           15.0.7      h0cdce71_0                    3.1 MiB    conda  llvm-openmp-15.0.7-h0cdce71_0.conda
markdown-it-py        3.0.0       pyhd8ed1ab_0                  62.8 KiB   conda  markdown-it-py-3.0.0-pyhd8ed1ab_0.conda
markupsafe            2.1.5       py311h459d7ec_0               26.9 KiB   conda  markupsafe-2.1.5-py311h459d7ec_0.conda
mdurl                 0.1.2       pyhd8ed1ab_0                  14.3 KiB   conda  mdurl-0.1.2-pyhd8ed1ab_0.conda
mkl                   2023.0.0    h84fe81f_26648                154 MiB    conda  mkl-2023.0.0-h84fe81f_26648.conda
mpc                   1.3.1       hfe3b2da_0                    113.6 KiB  conda  mpc-1.3.1-hfe3b2da_0.conda
mpfr                  4.2.1       h9458935_1                    628 KiB    conda  mpfr-4.2.1-h9458935_1.conda
mpmath                1.3.0       pyhd8ed1ab_0                  428.1 KiB  conda  mpmath-1.3.0-pyhd8ed1ab_0.conda
ncurses               6.5         h59595ed_0                    866.7 KiB  conda  ncurses-6.5-h59595ed_0.conda
networkx              3.3         pyhd8ed1ab_1                  1.1 MiB    conda  networkx-3.3-pyhd8ed1ab_1.conda
ocl-icd               2.3.2       hd590300_1                    132.5 KiB  conda  ocl-icd-2.3.2-hd590300_1.conda
openssl               3.3.1       h4bc722e_2                    2.8 MiB    conda  openssl-3.3.1-h4bc722e_2.conda
pydantic              2.8.2       pyhd8ed1ab_0                  285.7 KiB  conda  pydantic-2.8.2-pyhd8ed1ab_0.conda
pydantic-core         2.20.1      py311hb3a8bbb_0               1.5 MiB    conda  pydantic-core-2.20.1-py311hb3a8bbb_0.conda
pygments              2.18.0      pyhd8ed1ab_0                  858.7 KiB  conda  pygments-2.18.0-pyhd8ed1ab_0.conda
python                3.11.9      hb806964_0_cpython            29.5 MiB   conda  python-3.11.9-hb806964_0_cpython.conda
python-multipart      0.0.9       pyhd8ed1ab_0                  25.8 KiB   conda  python-multipart-0.0.9-pyhd8ed1ab_0.conda
python_abi            3.11        4_cp311                       6.2 KiB    conda  python_abi-3.11-4_cp311.conda
pytorch               2.4.0       py3.11_cuda12.4_cudnn9.1.0_0  1.3 GiB    conda  pytorch-2.4.0-py3.11_cuda12.4_cudnn9.1.0_0.tar.bz2
pytorch-cuda          12.4        hc786d27_6                    7 KiB      conda  pytorch-cuda-12.4-hc786d27_6.tar.bz2
pytorch-mutex         1.0         cuda                          2.8 KiB    conda  pytorch-mutex-1.0-cuda.tar.bz2
pyyaml                6.0.1       py311h459d7ec_1               195.9 KiB  conda  pyyaml-6.0.1-py311h459d7ec_1.conda
readline              8.2         h8228510_1                    274.9 KiB  conda  readline-8.2-h8228510_1.conda
retrieval_api         0.1.0                                                pypi   . (editable)
rich                  13.7.1      pyhd8ed1ab_0                  180 KiB    conda  rich-13.7.1-pyhd8ed1ab_0.conda
shellingham           1.5.4       pyhd8ed1ab_0                  14.2 KiB   conda  shellingham-1.5.4-pyhd8ed1ab_0.conda
sniffio               1.3.1       pyhd8ed1ab_0                  14.7 KiB   conda  sniffio-1.3.1-pyhd8ed1ab_0.conda
starlette             0.37.2      pyhd8ed1ab_0                  56.3 KiB   conda  starlette-0.37.2-pyhd8ed1ab_0.conda
sympy                 1.13.0      pypyh2585a3b_103              4.3 MiB    conda  sympy-1.13.0-pypyh2585a3b_103.conda
tbb                   2021.12.0   h434a139_3                    188.9 KiB  conda  tbb-2021.12.0-h434a139_3.conda
tk                    8.6.13      noxft_h4845f30_101            3.2 MiB    conda  tk-8.6.13-noxft_h4845f30_101.conda
torchtriton           3.0.0       py311                         233.7 MiB  conda  torchtriton-3.0.0-py311.tar.bz2
typer                 0.12.3      pyhd8ed1ab_0                  51.4 KiB   conda  typer-0.12.3-pyhd8ed1ab_0.conda
typer-slim            0.12.3      pyhd8ed1ab_0                  44.6 KiB   conda  typer-slim-0.12.3-pyhd8ed1ab_0.conda
typer-slim-standard   0.12.3      hd8ed1ab_0                    45 KiB     conda  typer-slim-standard-0.12.3-hd8ed1ab_0.conda
typing-extensions     4.12.2      hd8ed1ab_0                    9.9 KiB    conda  typing-extensions-4.12.2-hd8ed1ab_0.conda
typing_extensions     4.12.2      pyha770c72_0                  39 KiB     conda  typing_extensions-4.12.2-pyha770c72_0.conda
tzdata                2024a       h0c530f3_0                    117 KiB    conda  tzdata-2024a-h0c530f3_0.conda
uvicorn               0.30.3      py311h38be061_0               130.4 KiB  conda  uvicorn-0.30.3-py311h38be061_0.conda
xz                    5.2.6       h166bdaf_0                    408.6 KiB  conda  xz-5.2.6-h166bdaf_0.tar.bz2
```

BUT 

```
[tool.pixi.dependencies]
python = "3.11.*"
fastapi = ">=0.111.1,<0.112"
pytorch = { version = "2.4.0.*", channel = "pytorch" }
cpuonly = ">=2.0,<3"
```

Downloaded CPU only torch 

### Issue description

system requirements isn't working as expected. I haven't mentioned Cuda anywhere but still, the above toml install Cuda versions of pytorch working

### Expected behavior

Until cuda is mentioned it should download CPU versions only.

**Comment by ruben-arts** [comment]
Thanks for reporting! An even simpler repro is:
```
pixi init cuda-test && cd cuda-test
pixi add cuda-version
# this should fail but it doesn't
```
`cuda-version` depends on the `__cuda` virtual package and without telling pixi that that is available it should not solve for cuda.

**Comment by ruben-arts** [comment]
**TL;DR**: We're working on a fix. For now run `pixi add cpuonly` which excludes all `cuda` packages.

---

Together with @baszalmstra we've looked into this and the issue is as follows.

The `cuda-version` packages does not depend on the `__cuda >= 12.5` virtual package but adds it as a `constrains`. Which means, _if_ the `__cuda` package is in the solution it should be at least be version `12.5`. 

Since there are no dependencies on `__cuda` in this solution these constrains satisfy.  If _any_ package in your dependencies would depend on `__cuda`, this mechanism works as expected. 

In our solver ([resolvo](https://github.com/mamba-org/resolvo)) we need to add the understanding of what happens with the given virtual packages by pixi. As currently they're only added as optional packages and not as part of the solution. 

This issue has been present since we've been using `resolvo`, almost pixi's entire existence. This seems to be a recent issue as the cuda packages only define `constains` and no `dependencies` on the virtual package. (Which feels wrong to me but I don't know the underlying reasoning). 

We'll work on a solution in the coming days, meanwhile the easiest way to deal with it is to add the `cpuonly` package. Which avoids this whole mechanism.

**Comment by ruben-arts** [comment]
Fixed in: https://github.com/conda-incubator/rattler/pull/795

Will require a new release and bump in `pixi`.
